boot/boot_robustness_test.go

```diff
@@ -0,0 +1,304 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package boot_test
+
+import (
+	"fmt"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/boot/boottest"
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/snap"
+)
+
+// TODO:UC20: move this to bootloadertest package and use from i.e. managers_test.go ?
+func runBootloaderLogic(c *C, bl bootloader.Bootloader) (snap.PlaceInfo, error) {
+	// switch on which kind of bootloader we have
+	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
+	if ok {
+		return extractedRunKernelImageBootloaderLogic(c, ebl)
+	}
+
+	return pureenvBootloaderLogic(c, "kernel_status", bl)
+}
+
+// runBootloaderLogic implements the logic from the gadget snap bootloader,
+// namely that we transition kernel_status "try" -> "trying" and "trying" -> ""
+// and use try-kernel.efi when kernel_status is "try" and kernel.efi in all
+// other situations
+func extractedRunKernelImageBootloaderLogic(c *C, ebl bootloader.ExtractedRunKernelImageBootloader) (snap.PlaceInfo, error) {
+	m, err := ebl.GetBootVars("kernel_status")
+	c.Assert(err, IsNil)
+	kernStatus := m["kernel_status"]
+
+	kern, err := ebl.Kernel()
+	c.Assert(err, IsNil)
+	c.Assert(kern, Not(IsNil))
+
+	switch kernStatus {
+	case boot.DefaultStatus:
+	case boot.TryStatus:
+		// move to trying, use the try-kernel
```

Comment:
a remark, this is probably because we don't have  happy cases here,  but the code for this case here is never used in the tests afaict. maybe we should have happy tests with this setup, it's not a blocker though

Reply:
we could add some happy tests with this setup, it wouldn't too hard and would be a good thing to have tests for with the new implementation as well

Reply:
I'll queue this up for a followup PR